### PR TITLE
PIM-6737: Export - As Julia, I would like to export products models (csv, xlsx)

### PR DESCRIPTION
**Pull request by nidup**
@jjanvier we need to discuss about the product model reader. Here, I implement it with the repository findAll method as we do for families etc but i'd like use a proper cursor like for products, I don't know yet what are the impacts (needs a dedicated elastic search index for product models?).

As this PR is already pretty fat, I'd prefer pass this change with an upcoming dedicated PR implementing the index, cursor, etc thing + changing the product model reader implementation.

@ahocquard I introduced here the integration tests for product models export, the creation of product model, variant product, family variant, attribute option, etc is now usable in upcoming import/export tests to continue to replace legacy behat scenario by integration tests :wink: 

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | N
| Added integration tests           | Y
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php [hunk 1]
```diff
@@ -63,6 +66,37 @@ protected function createProduct(string $identifier, array $data = []) : Product
...
+    /**
+     * @param array  $data
+     *
+     * @return ProductModelInterface
+     */
+    protected function createProductModel(array $data = []) : ProductModelInterface
+    {
+        $productModel = $this->get('pim_catalog.factory.product_model')->create();
+        $this->get('pim_catalog.updater.product_model')->update($productModel, $data);
+        $this->get('pim_catalog.saver.product_model')->save($productModel);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: Shouldn't we validate prior to saving ?

File: src/Pim/Bundle/ConnectorBundle/Resources/config/job_constraints.yml [hunk 2]
```diff
@@ -4,6 +4,8 @@ parameters:
     pim_connector.job.job_parameters.constraint_collection_provider.simple_yaml_export.class:        Pim\Component\Connector\Job\JobParameters\ConstraintCollectionProvider\SimpleYamlExport
     pim_connector.job.job_parameters.constraint_collection_provider.product_csv_export.class:        Pim\Component\Connector\Job\JobParameters\ConstraintCollectionProvider\ProductCsvExport
     pim_connector.job.job_parameters.constraint_collection_provider.product_xlsx_export.class:       Pim\Component\Connector\Job\JobParameters\ConstraintCollectionProvider\ProductXlsxExport
+    pim_connector.job.job_parameters.constraint_collection_provider.product_model_csv_export.class:  Pim\Component\Connector\Job\JobParameters\ConstraintCollectionProvider\ProductModelCsvExport
+    pim_connector.job.job_parameters.constraint_collection_provider.product_model_xlsx_export.class: Pim\Component\Connector\Job\JobParameters\ConstraintCollectionProvider\ProductModelCsvExport
```
Comment (grena) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/job_constraints.yml]: Is it normal that it's the same class than `product_model_csv_export.class`?

File: src/Pim/Bundle/ConnectorBundle/Resources/config/job_defaults.yml [hunk 3]
```diff
@@ -4,6 +4,8 @@ parameters:
     pim_connector.job.job_parameters.default_values_provider.simple_yaml_export.class:        Pim\Component\Connector\Job\JobParameters\DefaultValuesProvider\SimpleYamlExport
     pim_connector.job.job_parameters.default_values_provider.product_csv_export.class:        Pim\Component\Connector\Job\JobParameters\DefaultValuesProvider\ProductCsvExport
     pim_connector.job.job_parameters.default_values_provider.product_xlsx_export.class:       Pim\Component\Connector\Job\JobParameters\DefaultValuesProvider\ProductXlsxExport
+    pim_connector.job.job_parameters.default_values_provider.product_model_csv_export.class:  Pim\Component\Connector\Job\JobParameters\DefaultValuesProvider\ProductModelCsvExport
+    pim_connector.job.job_parameters.default_values_provider.product_model_xlsx_export.class: Pim\Component\Connector\Job\JobParameters\DefaultValuesProvider\ProductModelCsvExport
```
Comment (grena) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/job_defaults.yml]: Same question regarding the use of the same class

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php (hunk 1 above)
Comment (jjanvier) [reply src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: yes please :)

File: src/Pim/Bundle/CatalogBundle/Resources/config/removers.yml [hunk 4]
```diff
@@ -26,7 +26,7 @@ services:
         arguments:
             - '@pim_catalog.object_manager.product'
             - '@event_dispatcher'
-            - 'Pim\Component\Catalog\Model\ProductInterface'
+            - '%pim_catalog.model.product.interface%'
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Resources/config/removers.yml]: :+1:

File: src/Pim/Component/Connector/Writer/File/Csv/ProductModelWriter.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            'fieldDelimiter' => $parameters->get('delimiter'),
+            'fieldEnclosure' => $parameters->get('enclosure'),
+            'shouldAddBOM'   => false,
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getItemIdentifier(array $product)
```
Comment (SamirBoulil) [inline src/Pim/Component/Connector/Writer/File/Csv/ProductModelWriter.php]: $productModel

File: src/Pim/Component/Connector/Writer/File/Xlsx/ProductModelWriter.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+     */
+    protected function getWriterConfiguration()
+    {
+        return ['type' => 'xlsx'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getItemIdentifier(array $product)
```
Comment (SamirBoulil) [inline src/Pim/Component/Connector/Writer/File/Xlsx/ProductModelWriter.php]: $productModel

File: src/Pim/Component/Connector/Job/JobParameters/ConstraintCollectionProvider/ProductModelCsvExport.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * Constraints for product model CSV export
+ *
+ * @author    Nicolas Dupont <nicolas@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductModelCsvExport implements ConstraintCollectionProviderInterface
+{
+    /** @var ConstraintCollectionProviderInterface */
+    protected $simpleProvider;
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Job/JobParameters/ConstraintCollectionProvider/ProductModelCsvExport.php]: I am not sure that properties should be protected, they should be private

Comment (aRn0D) [reply src/Pim/Component/Connector/Job/JobParameters/ConstraintCollectionProvider/ProductModelCsvExport.php]: That's the same for other class.

File: src/Pim/Component/Connector/Processor/Normalization/ProductModelProcessor.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        $this->stepExecution = $stepExecution;
+    }
+
+    /**
+     * Fetch medias on the local filesystem
+     *
+     * @param ProductModelInterface $productModel
+     * @param string                $directory
+     */
+    protected function fetchMedia(ProductModelInterface $productModel, $directory)
```
Comment (aRn0D) [inline src/Pim/Component/Connector/Processor/Normalization/ProductModelProcessor.php]: should be private

File: src/Pim/Component/Connector/spec/Job/JobParameters/ConstraintCollectionProvider/ProductModelCsvExportSpec.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+class ProductModelCsvExportSpec extends ObjectBehavior
+{
+    function let(ConstraintCollectionProviderInterface $decoratedProvider)
+    {
+        $this->beConstructedWith($decoratedProvider, ['my_supported_job_name']);
+    }
+
+    function it_is_a_provider()
+    {
+        $this->shouldImplement('Akeneo\Component\Batch\Job\JobParameters\ConstraintCollectionProviderInterface');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/spec/Job/JobParameters/ConstraintCollectionProvider/ProductModelCsvExportSpec.php]: ConstraintCollectionProviderInterface::class

File: src/Pim/Component/Connector/spec/Job/JobParameters/DefaultValuesProvider/ProductModelCsvExportSpec.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+{
+    function let(
+        DefaultValuesProviderInterface $decoratedProvider
+    ) {
+        $this->beConstructedWith($decoratedProvider, ['my_supported_job_name']);
+    }
+
+    function it_is_a_provider()
+    {
+        $this->shouldImplement('Akeneo\Component\Batch\Job\JobParameters\DefaultValuesProviderInterface');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/spec/Job/JobParameters/DefaultValuesProvider/ProductModelCsvExportSpec.php]: DefaultValuesProviderInterface::class

File: src/Pim/Component/Connector/spec/Processor/Normalization/ProductModelProcessorSpec.php [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+            $mediaFetcher,
+            $valuesFiller
+        );
+
+        $this->setStepExecution($stepExecution);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('\Pim\Component\Connector\Processor\Normalization\ProductModelProcessor');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/spec/Processor/Normalization/ProductModelProcessorSpec.php]: ProductModelProcessor::class

File: src/Pim/Component/Connector/spec/Processor/Normalization/ProductModelProcessorSpec.php [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    function it_is_an_item_processor()
+    {
+        $this->shouldImplement('\Akeneo\Component\Batch\Item\ItemProcessorInterface');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/spec/Processor/Normalization/ProductModelProcessorSpec.php]: ItemProcessorInterface::class

File: tests/catalog/technical/jobs.yml [hunk 13]
```diff
@@ -103,3 +103,13 @@ jobs:
...
+    csv_product_model_export:
+        connector: Akeneo CSV Connector
+        alias:     csv_product_model_export
+        label:     CSV product model export
+        type:      export
+        configuration:
+            delimiter:  ;
+            enclosure:  '"'
+            withHeader: true
+            filePath: /tmp/product_models.csv
```
Comment (grena) [inline tests/catalog/technical/jobs.yml]: Should we add the xlsx job too?

File: src/Pim/Component/Connector/spec/Reader/Database/ProductModelReaderSpec.php [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+class ProductModelReaderSpec extends ObjectBehavior
+{
+    function let(ProductModelRepositoryInterface $repository)
+    {
+        $this->beConstructedWith($repository);
+    }
+
+    function it_is_a_reader()
+    {
+        $this->shouldImplement('Akeneo\Component\Batch\Item\ItemReaderInterface');
```
Comment (aRn0D) [inline src/Pim/Component/Connector/spec/Reader/Database/ProductModelReaderSpec.php]: ::class

File: src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php [hunk 15]
```diff
@@ -34,6 +35,7 @@ public function normalize($productModel, $format = null, array $context = array(
 
         $data[self::FIELD_CODE] = $productModel->getCode();
         $data[self::FIELD_FAMILY_VARIANT] = $productModel->getFamilyVariant()->getCode();
+        $data[self::FIELD_PARENT] = null !== $productModel->getParent() ? $productModel->getParent()->getCode() : null;
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php]: Don't forget to modify the `Standard_format.md` specification file

File: src/Pim/Bundle/ConnectorBundle/Resources/config/job_constraints.yml (hunk 2 above)
Comment (nidup) [reply src/Pim/Bundle/ConnectorBundle/Resources/config/job_constraints.yml]: Yes, exact same code for now, i did like we did for the import. At least we'll be able to change the concrete implementation when we'll improve the product model export (for now it's a simple export + the "with media" config, later on we would use the export builder for instance). Other option would be to name it ProductModelFlatFileExport to cover csv & excel but we would need to change it with upcoming improvs.

File: src/Pim/Bundle/ConnectorBundle/Resources/config/job_defaults.yml (hunk 3 above)
Comment (nidup) [reply src/Pim/Bundle/ConnectorBundle/Resources/config/job_defaults.yml]: same

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php [hunk 16]
```diff
@@ -63,6 +66,37 @@ protected function createProduct(string $identifier, array $data = []) : Product
      *
      * @return ProductInterface
      */
+    protected function createVariantProduct(string $identifier, array $data = []) : ProductInterface
+    {
+        $product = $this->get('pim_catalog.builder.variant_product')->createProduct($identifier);
+        $this->get('pim_catalog.updater.product')->update($product, $data);
```
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: validate?

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php [hunk 17]
```diff
@@ -103,6 +151,20 @@ protected function createFamily(array $data = []) : FamilyInterface
...
 
     /**
+     * @param array $data
+     *
+     * @return FamilyVariantInterface
+     */
+    protected function createFamilyVariant(array $data = []) : FamilyVariantInterface
+    {
+        $family = $this->get('pim_catalog.factory.family_variant')->create();
+        $this->get('pim_catalog.updater.family_variant')->update($family, $data);
```
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: vaidate?

File: src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php (hunk 15 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php]: just FYI, now you use `??` http://php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce

File: src/Pim/Component/Connector/Reader/Database/ProductModelReader.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $this->repository = $repository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getResults()
+    {
+        // TODO PIM-6737: temporary until we discuss how to fetch all models, needs a new elastic search index
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/Database/ProductModelReader.php]: :+1: 
yes, imo, the idea was to use a dedicated index, as we do for products
we can talk IRL about it if you want

I totally agree we should do that in another PR :p

File: src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php (hunk 15 above)
Comment (nidup) [reply src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php]: fixed :+1:

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php (hunk 16 above)
Comment (nidup) [reply src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: fixed :+1:

File: src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php (hunk 17 above)
Comment (nidup) [reply src/Pim/Bundle/ConnectorBundle/tests/integration/Export/AbstractExportTestCase.php]: fixed :+1:

File: src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php (hunk 15 above)
Comment (nidup) [reply src/Pim/Component/Catalog/Normalizer/Standard/ProductModelNormalizer.php]: thx for hint, never used for now
